packages/core/src/modules/proofs/protocol/v2/V2ProofService.ts

```diff
@@ -0,0 +1,808 @@
+    this.protocolVersion = ProofProtocolVersion.V2
+    this.wallet = wallet
+    this.formatServiceMap = {
+      [PresentationRecordType.Indy]: indyProofFormatService,
+    }
+  }
+
+  public getVersion(): ProofProtocolVersion {
+    return this.protocolVersion
+  }
+
+  public async createProposal(
+    options: CreateProposalOptions
+  ): Promise<{ proofRecord: ProofRecord; message: AgentMessage }> {
+    const formats = []
+    for (const key of Object.keys(options.proofFormats)) {
+      const service = this.formatServiceMap[key]
+      formats.push(
+        await service.createRequest({
+          formats:
+            key === PresentationRecordType.Indy
+              ? await ProofsUtils.createRequestFromPreview(options)
+              : options.proofFormats,
+        })
+      )
+    }
+
+    const proposalMessage = new V2ProposalPresentationMessage({
+      attachmentInfo: formats,
+      comment: options.comment,
+      willConfirm: options.willConfirm,
+      goalCode: options.goalCode,
+    })
+
+    const proofRecord = new ProofRecord({
+      connectionId: options.connectionRecord.id,
+      threadId: proposalMessage.threadId,
+      state: ProofState.ProposalSent,
+      protocolVersion: ProofProtocolVersion.V2,
+    })
+
+    await this.proofRepository.save(proofRecord)
+
+    await this.didCommMessageRepository.saveOrUpdateAgentMessage({
+      agentMessage: proposalMessage,
+      role: DidCommMessageRole.Sender,
+      associatedRecordId: proofRecord.id,
+    })
+
+    this.eventEmitter.emit<ProofStateChangedEvent>({
+      type: ProofEventTypes.ProofStateChanged,
+      payload: { proofRecord, previousState: null },
+    })
+
+    return {
+      proofRecord: proofRecord,
+      message: proposalMessage,
+    }
+  }
+
+  public async createProposalAsResponse(
+    options: CreateProposalAsResponseOptions
+  ): Promise<{ proofRecord: ProofRecord; message: AgentMessage }> {
+    options.proofRecord.assertState(ProofState.RequestReceived)
+
+    const formats = []
+    for (const key of Object.keys(options.proofFormats)) {
+      const service = this.formatServiceMap[key]
+      formats.push(
+        await service.createProposal({
+          formats: options.proofFormats,
+        })
+      )
+    }
+
+    const proposalMessage = new V2ProposalPresentationMessage({
+      attachmentInfo: formats,
+      comment: options.comment,
+      goalCode: options.goalCode,
+      willConfirm: options.willConfirm,
+    })
+
+    await this.didCommMessageRepository.saveOrUpdateAgentMessage({
+      agentMessage: proposalMessage,
+      role: DidCommMessageRole.Sender,
+      associatedRecordId: options.proofRecord.id,
+    })
+
+    void this.updateState(options.proofRecord, ProofState.ProposalSent)
+
+    return { message: proposalMessage, proofRecord: options.proofRecord }
+  }
+
+  public async processProposal(messageContext: InboundMessageContext<AgentMessage>): Promise<ProofRecord> {
```

Comment:
This means you don't have to do `_proposalMessage as V2ProposalPresentationMessage`

```suggestion
  public async processProposal(messageContext: InboundMessageContext< V2ProposalPresentationMessage >): Promise<ProofRecord> {
```